diffmpm/element.py

```diff
@@ -1,30 +1,539 @@
+    def id_to_node_vel(self):
+        ...
 
-class Bar1D:
+    # @abc.abstractmethod
+    # def tree_flatten(self):
+    #     ...
+
+    # @classmethod
+    # @abc.abstractmethod
+    # def tree_unflatten(self):
+    #     ...
+
+
+class Linear1D(_Element):
     """
-    1D bar element with 2 nodes.
-
-    Attributes:
-
-    id: int
-        Element index.
-    
-    nid1: int
-        Node index 1.
-    
-    nid2: int
-        Node index 2.
-
-    length: float
-        Element length.
-    
-    particles: list
-        List of particles in bar.
+    Container for 1D line elements (and nodes).
+
+    Element ID:            0     1     2     3
+    Mesh:               +-----+-----+-----+-----+
+    Node IDs:           0     1     2     3     4
+
+    + : Nodes
+    +-----+ : An element
     """
 
-    def __init__(self):
-        self.id = None
-        self.node1 = None
-        self.node2 = None
-        self.size = jnp.array([0])
-        self.particles = []
\ No newline at end of file
+    def __init__(self, nelements: int, el_len: float):
+        """Initialize Linear1D.
+
+        Arguments
+        ---------
+        nelements : int
+            Number of elements.
+        el_len : float
+            Length of each element.
+        """
+        self.ids: jnp.ndarray = jnp.arange(nelements)
+        self.el_len: float = el_len
+        self.nodes: Nodes = Nodes(
+            nelements + 1, jnp.arange(nelements + 1).reshape(-1, 1, 1) * el_len
+        )
+
+    def id_to_node_ids(self, id: int):
+        """
+        Node IDs corresponding to element `id`.
+
+        Arguments
+        ---------
+        id : int
+            Element ID.
+
+        Returns
+        -------
+        jax.numpy.ndarray
+            Nodal IDs of the element. Shape of returned
+        array is (2, 1)
+        """
+        return jnp.array([id, id + 1]).reshape(2, 1)
+
+    def id_to_node_loc(self, id: int):
+        """
+        Node locations corresponding to element `id`.
+
+        Arguments
+        ---------
+        id : int
+            Element ID.
+
+        Returns
+        -------
+        jax.numpy.ndarray
+            Nodal locations for the element. Shape of returned
+        array is (len(id), 2, 1)
+        """
+        return self.nodes.loc[jnp.array([id, id + 1])].transpose(1, 0, 2, 3)
+
+    def id_to_node_vel(self, id: int):
+        """
+        Node velocities corresponding to element `id`.
+
+        Arguments
+        ---------
+        id : int
+            Element ID.
+
+        Returns
+        -------
+        jax.numpy.ndarray
+            Nodal velocities for the element. Shape of returned
+        array is (2, 1)
+        """
+        return self.nodes.velocity[jnp.array([id, id + 1])].reshape(2, 1)
+
+    def shapefn(self, xi):
+        """
+        Evaluate linear shape function.
+
+        Arguments
+        ---------
+        xi : float, array_like
+            Locations of particles in natural coordinates to evaluate
```

Comment:
natural coordinates specify which coordinates, the background mesh ones with nodes or the material points one.

Reply:
Natural coordinates are coordinates in the element, relative to the node. For example, if we're talking about 1D elements, say the mesh looks something like this:

+-------+-------+---o---+-------+-------+
0         1        2    p     3        4         5

Here, if p is the point right in the center of the element with nodes 2 and 3, then it's physical coordinates would be x=2.5. Now when the element is mapped to a standard element

+----o----+
-1    p     1

Then it's natural coordinate will be x=0. Does that make sense?

Reply:
yes, now its clear